### PR TITLE
Updated AsyncWorker (and defined AsyncProcessor) so that we get back …

### PR DESCRIPTION
**Pull request by josephcarroll**
…Futures for processed tasks. Those futures can then be used to block on and wait if desired.

Exceptions during async processing are now also caught and logged, rather than causing the thread to die and leaving us in a bad state.

### CODE REVIEW

**Comment by devgianlu** [comment]
I think there's still the issue that the task will be interrupted when the worker is closed. The problem originated from #155 where I need to send an event before shutting everything down, if it's not quick enough an exception will be thrown which is not a big deal because the request itself has been sent already. 

I was just wondering whether there was a way to get around it.

**Comment by josephcarroll** [comment]
Changed to a simpler Executor based approach. This means when you call close it will wait 1 minute for currently running tasks to complete.

File: api/src/main/resources/log4j.properties [hunk 1]
```diff
@@ -2,7 +2,7 @@ log4j.rootLogger=TRACE, stdout
 log4j.appender.stdout=org.apache.log4j.ConsoleAppender
 log4j.appender.stdout.Target=System.out
 log4j.appender.stdout.layout=org.apache.log4j.PatternLayout
-log4j.appender.stdout.layout.ConversionPattern=%d %-5p %c{1}:%L - %m%n
+log4j.appender.stdout.layout.ConversionPattern=%d [%t] %-5p %c{1}:%L - %m%n
```
Comment (devgianlu) [inline api/src/main/resources/log4j.properties:5]: What's this change?

Comment (josephcarroll) [reply api/src/main/resources/log4j.properties:5]: It logs the thread name [like-this]

**Pull request merged**